### PR TITLE
[#1497] Provide the AbstractCodeMiningTest abstract base class.

### PR DESCRIPTION
**Pull request by miklossy**
- Extend the 'Xtext UI Testing' test infrastructure by the
AbstractCodeMiningTest abstract class to provide a base infrastructure
for testing the code mining capabilities.
- Extend the Xtext Examples by CodeMiningTest test cases to
demonstrate the usage of the AbstractCodeMiningTest class.

Signed-off-by: Tamas Miklossy <miklossy@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		List<? extends ICodeMining> codeMinings;
+		try {
+			codeMinings = codeMiningProvider.provideCodeMinings(viewer, new NullProgressMonitor()).get();
+		} catch (InterruptedException | ExecutionException e) {
+			throw new RuntimeException(e);
+		}
+
+		int currentOffset = 0;
+
+		for (ICodeMining codeMining : codeMinings) {
```
Comment (cdietrich) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:82]: does this work with multiple codeminings? does it work with multiple codeminings at the same place? does it work if the codeminings are not sorted by offset?

Comment (miklossy) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:82]: Multiple codeminings are already supported, see org.eclipse.xtext.example.arithmetics.ui.tests.codemining.CodeMiningTest test class in the PR:
screenshot

I think, the other corner cases you mentioned are not supported, but the user has the possibility to override the `protected String insertCodeMinings(XtextEditor editor)` method to provide a custom implementation for such cases.

Comment (cdietrich) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:82]: i dont think we should let that up to the user ....

Comment (cdietrich) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:82]: the usual algorithm is to sort codeminings by offset backwards (apply them from bottom to top)

Comment (cdietrich) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:82]: e.g. adding 

```
		for (Property f :  EcoreUtil2.eAllOfType(resource.getContents().get(0), Property.class)) {
			int endOffset = NodeModelUtils.findActualNodeFor(f).getEndOffset();
			acceptor.accept(createNewLineContentCodeMining(endOffset, " Breaking Tamas Test"));
		}
```

to domain model code mining provider easily breaks your code

**Comment by miklossy** [comment]
I updated the PR by sorting the code mining according to their offset.

@cdietrich: Could you please review again?

**Comment by miklossy** [comment]
I updated the PR to also work for multiple codeminings at the same offset.

@cdietrich: Could you please review again?

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+			if (codeMining instanceof LineHeaderCodeMining) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	protected int getLastCodeMiningOnOffset(List<? extends ICodeMining> codeMinings, int offset) {
+		int i = 0;
+		for(;i<codeMinings.size();i++) {
```
Comment (cdietrich) [inline org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:157]: formatting

Comment (miklossy) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java]: done

File: org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java (hunk 1 above)
Comment (miklossy) [reply org.eclipse.xtext.ui.testing/src/org/eclipse/xtext/ui/testing/AbstractCodeMiningTest.java:86]: done
